### PR TITLE
Add GeoserveRegionsService to fetch regions from geoserve

### PR DESCRIPTION
**Pull request by ehunter-usgs**
This is part of hazdev-realtime (issue 24)

### CODE REVIEW

File: src/main/java/gov/usgs/earthquake/origin/OriginIndexerModule.java [hunk 1]
```diff
@@ -92,16 +121,116 @@ public void setPlacesService(GeoservePlaces geoservePlaces) {
...
+    this.distanceThreshold = Double.parseDouble(
+        config.getProperty(
+            GEOSERVE_DISTANCE_THRESHOLD_PROPERTY,
+            DEFAULT_GEOSERVE_DISTANCE_THRESHOLD.toString()
+        )
+    );
 
     LOGGER.config(String.format("[%s] GeoservePlacesService(%s, %d, %d)", this.getName(), endpointUrl, connectTimeout,
         readTimeout));
     this.setPlacesService(new GeoservePlacesService(endpointUrl, connectTimeout, readTimeout));
```
Comment (jmfee-usgs) [inline src/main/java/gov/usgs/earthquake/origin/OriginIndexerModule.java:171]: I think you also need to configure the GeoserveRegionsService here?

File: src/main/java/gov/usgs/earthquake/origin/OriginIndexerModule.java [hunk 2]
```diff
@@ -26,26 +31,50 @@
...
 
   private GeoservePlaces geoservePlaces;
+  private GeoserveRegionsService geoserveRegions;
 
   public static final String ENDPOINT_URL_PROPERTY = "endpointUrl";
   public static final String CONNECT_TIMEOUT_PROPERTY = "connectTimeout";
   public static final String READ_TIMEOUT_PROPERTY = "readTimeout";
+  public static final String GEOSERVE_DISTANCE_THRESHOLD_PROPERTY = "geoserveDistanceThreshold";
+
+  public static final Double DEFAULT_GEOSERVE_DISTANCE_THRESHOLD = 300.0;
```
[the remark this answers is not in the document]
Comment (emartinez-usgs) [reply src/main/java/gov/usgs/earthquake/origin/OriginIndexerModule.java:41]: an integer may do the trick here. they are more performant than dealing with doubles. I can't see where we would need the precision to say 300.1 km and the earth isn't large enough to require a long.

File: src/main/java/gov/usgs/earthquake/geoserve/GeoservePlaces.java [hunk 3]
```diff
@@ -4,6 +4,8 @@
 import java.math.BigDecimal;
 import java.net.MalformedURLException;
 
+import javax.json.JsonObject;
+
 public interface GeoservePlaces {
```
[the remark this answers is not in the document]
Comment (ehunter-usgs) [reply src/main/java/gov/usgs/earthquake/geoserve/GeoservePlaces.java:9]: This is the return type for getNearestPlace

[the remark this answers is not in the document]
Comment (ehunter-usgs) [reply src/main/java/gov/usgs/earthquake/geoserve/GeoservePlaces.java:9]: oh my bad, I misread this... I agree. We don't really need the Interface. Is that a java convention?

File: src/main/java/gov/usgs/earthquake/origin/OriginIndexerModule.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (ehunter-usgs) [reply src/main/java/gov/usgs/earthquake/origin/OriginIndexerModule.java]: Now uses Integer.

File: src/main/java/gov/usgs/earthquake/geoserve/GeoservePlacesService.java [hunk 4]
```diff
@@ -105,12 +64,12 @@ public JsonObject getEventPlaces(BigDecimal latitude, BigDecimal longitude)
     }
   }
 
-  public String getEventTitle(BigDecimal latitude, BigDecimal longitude) throws IOException, MalformedURLException {
+  public JsonObject getNearestPlace(BigDecimal latitude, BigDecimal longitude) throws IOException, MalformedURLException {
     JsonObject places = this.getEventPlaces(latitude, longitude);
     JsonArray features = places.getJsonArray("features");
     JsonObject feature = features.get(0).asJsonObject();
```
[the remark this answers is not in the document]
Comment (ehunter-usgs) [reply src/main/java/gov/usgs/earthquake/geoserve/GeoservePlacesService.java]: Updated.

File: src/main/java/gov/usgs/earthquake/geoserve/GeoservePlaces.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (emartinez-usgs) [reply src/main/java/gov/usgs/earthquake/geoserve/GeoservePlaces.java:9]: I think I created it thinking we might have lots of "Places" type service things, but it turns out no, we don't.

File: src/main/java/gov/usgs/earthquake/geoserve/GeoserveRegionsService.java [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+    return this.connectTimeout;
+  }
+
+  public String getEndpointURL() {
+    return this.endpointUrl;
+  }
+
+  public JsonObject getFeRegion(BigDecimal latitude, BigDecimal longitude)
+      throws IOException, MalformedURLException {
+    final URL url = new URL(String.format("%s?type=fe&latitude=%s&longitude=%s", this.endpointUrl,
```
[the remark this answers is not in the document]
Comment (ehunter-usgs) [reply src/main/java/gov/usgs/earthquake/geoserve/GeoserveRegionsService.java]: Now uses string concatenation, I also think it reads easier now.

File: src/main/java/gov/usgs/earthquake/geoserve/GeoserveRegionsService.java [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+  public GeoserveRegionsService(final String endpointUrl) {
+    this(endpointUrl, DEFAULT_CONNECT_TIMEOUT, DEFAULT_READ_TIMEOUT);
+  }
+
+  public GeoserveRegionsService(final int connectTimeout, final int readTimeout) {
+    this(DEFAULT_ENDPOINT_URL, connectTimeout, readTimeout);
+  }
+
+  public GeoserveRegionsService(final String endpointUrl, final int connectTimeout, final int readTimeout) {
+    this.endpointUrl = endpointUrl;
```
[the remark this answers is not in the document]
Comment (ehunter-usgs) [reply src/main/java/gov/usgs/earthquake/geoserve/GeoserveRegionsService.java]: Updated.

File: src/main/java/gov/usgs/earthquake/origin/OriginIndexerModule.java (hunk 1 above)
Comment (ehunter-usgs) [reply src/main/java/gov/usgs/earthquake/origin/OriginIndexerModule.java]: Updated.

File: src/main/java/gov/usgs/earthquake/geoserve/GeoservePlaces.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (ehunter-usgs) [reply src/main/java/gov/usgs/earthquake/geoserve/GeoservePlaces.java]: Updated.

File: src/test/java/gov/usgs/earthquake/origin/OriginIndexerModuleTest.java [hunk 7]
```diff
@@ -89,13 +100,83 @@ public void getProductSummary() throws Exception {
...
+                .add("admin1_name", "admin1_name").add("country_name", "country_name").add("distance", distance)
+                .add("azimuth", azimuth))
+        .build();
+    Assert.assertEquals(expectation, module.formatEventTitle(feature));
+
+    // US location
+    expectation = String.format("0 km S of name, admin1_name");
+    feature = Json.createObjectBuilder()
+        .add("properties",
+            Json.createObjectBuilder().add("name", "name").add("country_code", "us").add("admin1_name", "admin1_name")
```
Comment (jmfee-usgs) [inline src/test/java/gov/usgs/earthquake/origin/OriginIndexerModuleTest.java:126]: Minor formatting, put each add on its own line and alphabetize unless another order is needed

File: src/main/java/gov/usgs/earthquake/origin/OriginIndexerModule.java [hunk 8]
```diff
@@ -83,25 +112,175 @@ public int getSupportLevel(Product product) {
...
+        new GeoservePlacesService(
+          placesEndpointUrl,
+          placesEndpointConnectTimeout,
+          placesEndpointReadTimeout
+        )
+    );
+
+    // Geoserve Regions Endpoint configuration
+    String regionsEndpointUrl = config.getProperty(
+        ENDPOINT_URL_PROPERTY,
```
[the remark this answers is not in the document]
Comment (ehunter-usgs) [reply src/main/java/gov/usgs/earthquake/origin/OriginIndexerModule.java]: Updated.

File: src/main/java/gov/usgs/earthquake/origin/OriginIndexerModule.java [hunk 9]
```diff
@@ -83,25 +112,175 @@ public int getSupportLevel(Product product) {
...
+   * place, or if the nearest place is outside of the distance threshold
+   * return the fe region name
+   *
+   * @param latitude event latitude in degrees
+   * @param longitude event longitude in degrees
+   *
+   * @return {String} event name
+   */
+  public String getEventTitle(BigDecimal latitude, BigDecimal longitude) throws IOException {
+    JsonObject feature = this.geoservePlaces.getNearestPlace(latitude, longitude);
```
[the remark this answers is not in the document]
Comment (ehunter-usgs) [reply src/main/java/gov/usgs/earthquake/origin/OriginIndexerModule.java]: Updated.

File: src/test/java/gov/usgs/earthquake/origin/OriginIndexerModuleTest.java (hunk 7 above)
Comment (ehunter-usgs) [reply src/test/java/gov/usgs/earthquake/origin/OriginIndexerModuleTest.java]: Updated.

**Pull request merged**